sympy/stats/stochastic_process_types.py

```diff
@@ -0,0 +1,230 @@
+
+    def probability(self, condition):
+        raise NotImplementedError()
+
+class DiscreteMarkovChain(StochasticProcess):
+    """
+    Represents discrete Markov chain.
+
+    Parameters
+    ==========
+
+    sym: Symbol
+    state_space: Set
+        Optional, by default, S.Reals
+    trans_probs: Matrix/ImmutableMatrix/MatrixSymbol
+        Optional, by default, None
+
+    Examples
+    ========
+
+    >>> from sympy.stats.stochastic_process_types import DiscreteMarkovChain
+    >>> from sympy import Matrix, MatrixSymbol, Eq
+    >>> from sympy.stats import P
+    >>> T = Matrix([[0.5, 0.2, 0.3],[0.2, 0.5, 0.3],[0.2, 0.3, 0.5]])
+    >>> Y = DiscreteMarkovChain("Y", [0, 1, 2], T)
+    >>> YS = DiscreteMarkovChain("Y")
+    >>> Y.state_space
+    {0, 1, 2}
+    >>> Y.trans_probs
+    Matrix([
+    [0.5, 0.2, 0.3],
+    [0.2, 0.5, 0.3],
+    [0.2, 0.3, 0.5]])
+    >>> TS = MatrixSymbol('T', 3, 3)
+    >>> P(Eq(YS[3], 2), Eq(YS[1], 1), trans_probs=TS)
+    T[0, 2]*T[1, 0] + T[1, 1]*T[1, 2] + T[1, 2]*T[2, 2]
+    >>> P(Eq(Y[3], 2), Eq(Y[1], 1)).round(2)
+    0.36
+    """
+
+    is_Discrete = True
+    is_Continuous = False
+
+    index_set = S.Naturals0
+
+    def __new__(cls, sym, state_space=S.Reals, trans_probs=None):
+        if isinstance(sym, string_types):
+            sym = Symbol(sym)
+        if not isinstance(sym, Symbol):
+            raise TypeError("Name of stochastic process should be either only "
+                            "a string or Symbol.")
+        if isinstance(state_space, set):
+            state_space = list(state_space)
+        if isinstance(state_space, (list, tuple)):
+            state_space = FiniteSet(*state_space)
+        if not isinstance(state_space, Set):
+            raise TypeError("state_space should be an instance of Set/list/tuple/set only.")
+        if trans_probs != None:
+            if not isinstance(trans_probs, (Matrix, MatrixSymbol, ImmutableMatrix)):
+                raise TypeError("Transition probabilities etiher should "
+                                "be a Matrix or a MatrixSymbol.")
+            if isinstance(trans_probs, Matrix):
+                trans_probs = ImmutableMatrix(trans_probs.tolist())
+        return Basic.__new__(cls, sym, state_space, trans_probs)
+
+    @property
+    def trans_probs(self):
+        """
+        Transition probabilities of discrete Markov chain,
+        either an instance of Matrix or MatrixSymbol.
+        """
+        return self.args[2]
+
+    def __getitem__(self, time):
+        """
+        For indexing discrete Markov chain.
+
+        Returns
+        =======
+
+        RandomIndexedSymbol
+        """
+        if time not in self.index_set:
+            raise IndexError("%s is not in the index set of %s"%(time, self.symbol))
+        idx_obj = Indexed(self.symbol, time)
+        pspace_obj = StochasticPSpace(self.symbol, self)
+        return RandomIndexedSymbol(idx_obj, pspace_obj)
```

Comment:
This looks like it would apply to any discrete time stochastic process. Maybe it should be moved to a new superclass DiscreteTimeStochasticProcess...